OpenScienceJournal/whistlepunk_library/src/main/java/com/google/android/apps/forscience/whistlepunk/RecordFragment.java

```diff
@@ -1185,8 +1185,8 @@ private void loadIncludedSensors(final List<GoosciSensorLayout.SensorLayout> lay
                             "add external sensors") {
                         @Override
                         public void success(ExperimentSensors sensors) {
-                            mSensorRegistry.setExcludedIds(sensors.getExcludedSensorIds());
-                            updateExternalSensors(sensors.getIncludedSensors());
+                            mSensorRegistry.setExcludedIds(sensors.getExcludedInternalSensorIds());
```

Comment:
I don't understand this. Can you explain why this is being changed?

Reply:
This is basically reflecting changes I've done on how experiment sensors are retrieved from the database and how the ExperimentSensors class brings that list from the database to the rest of the code.
Before the change, a straight query was extracting sensors of the experiment from a database table, along with an included/not-included flag. External sensor further info were extracted with nested query from another table. It is a little bit odd how it works.
Internal sensors are always all included in the experiment, unless they have a record in the table with the not-included flag; external sensors are not included by default, unless they have a record in the table with the included flag.
I kept this logic, in order to not affect the data model, but I rewrote the query in order to have only one with an outer join between the experiment sensor table and the external sensor details one. I changed the fact that when an external device is added to an experiment, all its sensors are written in the experiment sensor table, by default with the "not included" flag. When the user adds or removes an external sensor, its corresponding record "included" flag is reflecting the change.
Before that change, ExperimentSensors was adopting this model:

List<ConnectableSensor> includedSensors
Set<String> excludedSensorIds

This was not satisfying my needs because a disabled external sensor, in this scenario, has no ConnectableSensor instance associated, so I cannot retrieve its details (icon, name, etc) without running further queries and changing a lot of code in the UI part. Thus, I've changed the ExperimentSensors structure to:

List<ConnectableSensor> externalSensors
Set<String> excludedInternalSensorIds

External sensor, in this scenario, are always all loaded, and their description, title, etc are always available in the UI layer. Wether they are selected for recording, is reported in the "mIncluded" boolean flag of the ConnectableSensor class. The exclusion list, at this point, is only about internal sensors, that doesn't have the need to have always an associated ConnectableSensor instance, since their key information are always directly available without the need to query a database or an external device. This makes the external sensors handling in the app smarter, quicker and cleaner.